ReactiveCocoa/Swift/Property.swift

```diff
@@ -8,66 +8,457 @@ public protocol PropertyType {
 	/// The current value of the property.
 	var value: Value { get }
 
-	/// A producer for Signals that will send the property's current value,
-	/// followed by all changes over time.
+	/// A producer for signals that sends the property's current value,
+	/// followed by all changes over time. It completes when the property
+	/// has deinitialized, or has no further change.
 	var producer: SignalProducer<Value, NoError> { get }
 
-	/// A signal that will send the property's changes over time.
+	/// A signal that will send the property's changes over time. It
+	/// completes when the property has deinitialized, or has no further
+	/// change.
 	var signal: Signal<Value, NoError> { get }
 }
 
-/// A read-only property that allows observation of its changes.
+/// Represents an observable property that can be mutated directly.
+///
+/// Only classes can conform to this protocol, because instances must support
+/// weak references (and value types currently do not).
+public protocol MutablePropertyType: class, PropertyType {
+	/// The current value of the property.
+	var value: Value { get set }
+}
+
+/// Protocol composition operators
+///
+/// The producer and the signal of transformed properties would complete
+/// only when its source properties have deinitialized.
+///
+/// A transformed property would retain its ultimate source, but not
+/// any intermediate property during the composition.
+extension PropertyType {
+	/// Lifts an unary SignalProducer operator to operate upon PropertyType instead.
+	@warn_unused_result(message="Did you forget to use the composed property?")
+	private func lift<U>(@noescape transform: SignalProducer<Value, NoError> -> SignalProducer<U, NoError>) -> AnyProperty<U> {
+		return AnyProperty(transforming: self, using: transform)
+	}
+
+	/// Lifts a binary SignalProducer operator to operate upon PropertyType instead.
+	@warn_unused_result(message="Did you forget to use the composed property?")
+	private func lift<P: PropertyType, U>(transform: SignalProducer<Value, NoError> -> SignalProducer<P.Value, NoError> -> SignalProducer<U, NoError>) -> P -> AnyProperty<U> {
+		return { otherProperty in
+			return AnyProperty(transformingFirst: self, second: otherProperty, using: transform)
+		}
+	}
+
+	/// Maps the current value and all subsequent values to a new property.
+	@warn_unused_result(message="Did you forget to use the composed property?")
+	public func map<U>(transform: Value -> U) -> AnyProperty<U> {
+		return lift { $0.map(transform) }
+	}
+
+	/// Combines the current value and the subsequent values of two `Property`s in
+	/// the manner described by `Signal.combineLatestWith:`.
+	@warn_unused_result(message="Did you forget to use the composed property?")
+	public func combineLatest<P: PropertyType>(with other: P) -> AnyProperty<(Value, P.Value)> {
```

Comment:
We should call this `combineLatestWith` for now for consistency. It can be renamed in RAC5 when we Swift3-ify all our functions.